src/EditorFeatures/Core/GoToNextAndPreviousMethodCommandHandler.cs

```diff
@@ -0,0 +1,155 @@
+            }
+
+            var caretPoint = textView.GetCaretPoint(subjectBuffer);
+            if (!caretPoint.HasValue)
+            {
+                nextHandler();
+                return;
+            }
+
+            int? targetPosition = null;
+            var waitResult = _waitIndicator.Wait("", allowCancel: true, action: waitContext =>
+            {
+                targetPosition = GetTargetPosition(document, caretPoint.Value.Position, next, waitContext.CancellationToken);
+            });
+
+            if (waitResult == WaitIndicatorResult.Canceled || targetPosition == null)
+            {
+                return;
+            }
+
+            var targetLocation = new SnapshotPoint(subjectBuffer.CurrentSnapshot, targetPosition.Value);
+            var viewLocation = textView.BufferGraph.MapUpToBuffer(targetLocation, PointTrackingMode.Positive, PositionAffinity.Successor, textView.TextBuffer);
+            if (viewLocation.HasValue)
+            {
+                textView.Caret.MoveTo(viewLocation.Value);
+            }
+        }
+
+        /// <summary>
+        /// Internal for testing purposes.
+        /// </summary>
+        internal static int? GetTargetPosition(Document document, int caretPosition, bool next, CancellationToken cancellationToken)
+        {
+            var syntaxFactsService = document.GetLanguageService<ISyntaxFactsService>();
+            if (syntaxFactsService == null)
+            {
+                return null;
+            }
+
+            var root = document.GetSyntaxRootAsync(cancellationToken).WaitAndGetResult(cancellationToken);
+            var currentMember = GetMember(syntaxFactsService, caretPosition, root);
+            if (currentMember == null)
+            {
+                return null;
+            }
+
+            var members = syntaxFactsService.GetMethodLevelMembers(root);
+            var index = members.IndexOf(currentMember);
+            if (index < 0)
+            {
+                Debug.Fail("Couldn't find current member in members?");
+                return null;
+            }
+
+            if (next)
+            {
+                // If we're in leading trivia, we just want to go to the start of this member.
+                if (caretPosition >= currentMember.Span.Start)
+                {
+                    index++;
+                    if (index == members.Count)
+                    {
+                        index = 0;
+                    }
+                }
+            }
+            else
+            {
+                // If we're in trailing trivia, we just want to go to the start of this member.
+                if (caretPosition <= currentMember.Span.End)
+                {
+                    index--;
+                    if (index < 0)
+                    {
+                        index = members.Count - 1;
+                    }
+                }
+            }
+
+            // TODO: Better position within the node (e.g. attributes?)
```

Comment:
Interestingly, VS 2013 does *exactly* the same thing (goes to the beginning of the attributes with no selection).